jax/_src/cloud_tpu_init.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+def cloud_tpu_init():
+  """Automatically sets Cloud TPU topology env vars.
+
+  **This must be called before the TPU runtime is loaded, which happens as soon
+  as JAX's C++ backend is loaded! I.e. call this before xla_bridge or xla_client
+  is imported.**
+
+  These environment variables are used to tell the TPU runtime what kind of mesh
+  topology to use. It assumes a single-host topology by default, so we manually
+  set them here to default to the full pod slice if applicable.
+
+  This will not set any env vars if a single topology-related env var is already
+  set.
+  """
+
+  if not _running_in_cloud_tpu_vm():
+    return
+
+  # If the user has set any topology-related env vars, don't set any
+  # automatically.
+  if any([
+      os.environ.get("CLOUD_TPU_TASK_ID", None),
+      os.environ.get("TPU_CHIPS_PER_HOST_BOUNDS", None),
+      os.environ.get("TPU_HOST_BOUNDS", None),
+      os.environ.get("TPU_MESH_CONTROLLER_ADDRESS", None),
+      os.environ.get("TPU_MESH_CONTROLLER_PORT", None),
+      os.environ.get("TPU_VISIBLE_DEVICES", None),
+  ]):
+    return
+
+  # pylint: disable=import-outside-toplevel
+  # Don't assume non-Cloud TPU environments have requests installed
+  import requests
+  # pylint: enable=import-outside-toplevel
+  def get_metadata(key):
+    return requests.get(
+        'http://metadata.google.internal/computeMetadata'
```

Comment:
This URL is configurable in TF-TPU because Shawn wanted to do something weird; maybe we could recognize the same env var: https://github.com/tensorflow/tensorflow/pull/40317

Reply:
Done